File: src/features/import/utils/formatDetector.ts
import JSZip from 'jszip';
import { ExportFormat } from '@/features/export/utils/formatMapping';
import { ProjectType } from '@/lib/db';

export interface DetectionResult {
  format: ExportFormat;
  projectType: ProjectType;
  confidence: number; // 0-1
  classCount?: number;
}

export async function detectFormat(zip: JSZip): Promise<DetectionResult> {
  const files = Object.keys(zip.files);
  const lowerFiles = files.map((file) => file.toLowerCase());
  
  // Detect YOLO format
  if (hasFile(lowerFiles, 'classes.txt') && hasFile(lowerFiles, 'data.yaml')) {
    const yoloResult = await detectYOLOFormat(zip, lowerFiles);
    if (yoloResult) return yoloResult;
  }

  // Detect U-Net Masks format first when masks/images folders are present
  if (hasFolder(lowerFiles, 'masks') && hasFolder(lowerFiles, 'images')) {
    const unetResult = await detectUNetFormat(zip, lowerFiles);
    if (unetResult) return unetResult;
  }

  // Detect COCO format
  if (hasFile(lowerFiles, 'annotations.json')) {
    const cocoResult = await detectCOCOFormat(zip, lowerFiles);
    if (cocoResult) return cocoResult;
  }

  // Detect TIX (Annotix) format: annotations.json + images/ folder with our schema
  if (hasFile(lowerFiles, 'annotations.json') && hasFolder(lowerFiles, 'images')) {
    const tixResult = await detectTIXFormat(zip, lowerFiles);
    if (tixResult) return tixResult;
  }

  // Detect Pascal VOC format
  if (hasFolder(lowerFiles, 'annotations') && hasFolder(lowerFiles, 'images')) {
    const vocResult = await detectPascalVOCFormat(zip, lowerFiles);
    if (vocResult) return vocResult;
  }

  // Detect CSV format
  if (hasFile(lowerFiles, 'annotations.csv')) {
    const csvResult = await detectCSVFormat(zip, lowerFiles);
    if (csvResult) return csvResult;
  }

  // Detect Folders by Class format
  const folderResult = await detectFoldersByClassFormat(zip, lowerFiles);
  if (folderResult) return folderResult;

  throw new Error('Unable to detect dataset format');
}

async function detectYOLOFormat(zip: JSZip, files: string[]): Promise<DetectionResult | null> {
  try {
    const classesFile = zip.file('classes.txt');
    if (!classesFile) return null;

    const classesContent = await classesFile.async('text');
    const classCount = classesContent.trim().split('\n').length;

    // Check if segmentation (has polygon format in .txt files)
    const txtFiles = files.filter(f => f.startsWith('labels/') && f.endsWith('.txt'));
    
    if (txtFiles.length === 0) return null;

    const firstTxtFile = zip.file(txtFiles[0]);
    if (!firstTxtFile) return null;

    const content = await firstTxtFile.async('text');
    const isSegmentation = detectSegmentationFormat(content);

    return {
      format: isSegmentation ? 'yolo-segmentation' : 'yolo-detection',
      projectType: isSegmentation ? 'polygon' : 'bbox',
      confidence: 0.95,
      classCount,
    };
  } catch {
    return null;
  }
}

async function detectCOCOFormat(zip: JSZip, files: string[]): Promise<DetectionResult | null> {
  try {
    const annotFile = zip.file('annotations.json');
    if (!annotFile) return null;

    const content = await annotFile.async('text');
    const data = JSON.parse(content);

    if (!data.annotations || !data.images || !data.categories) return null;

    // Determine if it's instance segmentation or bbox
    const hasSegmentation = data.annotations.some((a: any) => a.segmentation);
    const projectType = hasSegmentation ? 'instance-segmentation' : 'bbox';

    return {
      format: 'coco',
      projectType,
      confidence: 0.95,
      classCount: data.categories.length,
    };
  } catch {
    return null;
  }
}

async function detectPascalVOCFormat(zip: JSZip, files: string[]): Promise<DetectionResult | null> {
  try {
    const xmlFiles = files.filter(f => f.startsWith('Annotations/') && f.endsWith('.xml'));
    
    if (xmlFiles.length === 0) return null;

    const firstXmlFile = zip.file(xmlFiles[0]);
    if (!firstXmlFile) return null;

    const content = await firstXmlFile.async('text');
    
    // Basic validation - check for Pascal VOC XML structure
    if (!content.includes('<annotation>') || !content.includes('<object>')) {
      return null;
    }

    const classCount = extractPascalVOCClasses(zip, files).length;

    return {
      format: 'pascal-voc',
      projectType: 'bbox',
      confidence: 0.9,
      classCount,
    };
  } catch {
    return null;
  }
}

async function detectCSVFormat(zip: JSZip, files: string[]): Promise<DetectionResult | null> {
  try {
    const csvFile = zip.file('annotations.csv');
    if (!csvFile) return null;

    const content = await csvFile.async('text');
    const lines = content.trim().split('\n');
    
    if (lines.length < 2) return null;

    const header = lines[0].toLowerCase();
    const classesFile = zip.file('classes.csv');
    
    let classCount = 0;
    if (classesFile) {
      const classesContent = await classesFile.async('text');
      classCount = classesContent.trim().split('\n').length - 1; // -1 for header
    }

    // Detect CSV type
    if (header.includes('x,') && header.includes('y,') && header.includes('width,') && header.includes('height,')) {
      return {
        format: 'csv-detection',
        projectType: 'bbox',
        confidence: 0.9,
        classCount,
      };
    }

    if (header.includes('keypoint')) {
      return {
        format: 'csv-keypoints',
        projectType: 'keypoints',
        confidence: 0.9,
        classCount,
      };
    }

    if (header.includes('landmark')) {
      return {
        format: 'csv-landmarks',
        projectType: 'landmarks',
        confidence: 0.9,
        classCount,
      };
    }

    if (header.includes('label') || header.includes('class')) {
      return {
        format: 'csv-classification',
        projectType: 'classification',
        confidence: 0.85,
        classCount,
      };
    }

    return null;
  } catch {
    return null;
  }
}

async function detectUNetFormat(zip: JSZip, files: string[]): Promise<DetectionResult | null> {
  try {
    const maskFiles = files.filter((file) => {
      if (!file.startsWith('masks/')) return false;
      return ['.png', '.jpg', '.jpeg', '.bmp', '.webp'].some((ext) => file.endsWith(ext));
    });
    
    if (maskFiles.length === 0) return null;

    return {
      format: 'unet-masks',
      projectType: 'instance-segmentation',
      confidence: 0.9,
      classCount: 2, // Binary masks
    };
  } catch {
    return null;
  }
}

async function detectTIXFormat(zip: JSZip, files: string[]): Promise<DetectionResult | null> {
  try {
    const annotFile = zip.file('annotations.json');
    if (!annotFile) return null;

    const content = await annotFile.async('text');
    const data = JSON.parse(content);

    // Expecting a structure like { project, classes, images }
    if (!data.images || !Array.isArray(data.images)) return null;

    const mapProjectType = (type: string | undefined): ProjectType | null => {
      if (!type) return null;

      const normalized = String(type).toLowerCase();
      const mapped: Record<string, ProjectType> = {
        bbox: 'bbox',
        detection: 'bbox',
        mask: 'mask',
        segmentation: 'mask',
        polygon: 'polygon',
        keypoint: 'keypoints',
        keypoints: 'keypoints',
        landmark: 'landmarks',
        landmarks: 'landmarks',
        obb: 'obb',
        classification: 'classification',
        multilabel: 'multi-label-classification',
        'multi-label-classification': 'multi-label-classification',
        instanceseg: 'instance-segmentation',
        instancesegmentation: 'instance-segmentation',
        'instance-segmentation': 'instance-segmentation',
      };

      return mapped[normalized] || null;
    };

    const projectClasses = Array.isArray(data.project?.classes)
      ? data.project.classes
      : Array.isArray(data.classes)
        ? data.classes
        : [];
    const classCount = projectClasses.length;

    // First choice: explicit project.type from file
    let projectType: ProjectType = mapProjectType(data.project?.type) || 'bbox';

    // Fallback inference from annotation types when needed
    if (projectType === 'bbox') {
      for (const img of data.images) {
        const anns = Array.isArray(img.annotations) ? img.annotations : [];
        for (const ann of anns) {
          const annType = mapProjectType(ann?.type);
          if (annType && annType !== 'bbox') {
            projectType = annType;
            break;
          }

          if (ann?.segmentation) {
            projectType = 'instance-segmentation';
            break;
          }
        }

        if (projectType !== 'bbox') break;
      }
    }

    return {
      format: 'tix',
      projectType,
      confidence: 0.95,
      classCount,
    };
  } catch {
    return null;
  }
}

async function detectFoldersByClassFormat(zip: JSZip, files: string[]): Promise<DetectionResult | null> {
  try {
    // Look for image extensions in first level folders
    const imageFolders = new Set<string>();
    const imageExtensions = ['.jpg', '.jpeg', '.png', '.gif', '.bmp'];

    files.forEach(file => {
      const parts = file.split('/');
      if (parts.length === 2) {
        const ext = parts[1].toLowerCase();
        if (imageExtensions.some(e => ext.endsWith(e))) {
          imageFolders.add(parts[0]);
        }
      }
    });

    if (imageFolders.size >= 2) {
      return {
        format: 'folders-by-class',
        projectType: 'classification',
        confidence: 0.85,
        classCount: imageFolders.size,
      };
    }

    return null;
  } catch {
    return null;
  }
}

function hasFile(files: string[], filename: string): boolean {
  const target = filename.toLowerCase();
  return files.some((file) => {
    const normalized = file.toLowerCase();
    return (normalized.endsWith(target) && !normalized.includes('/')) || normalized === target;
  });
}

function hasFolder(files: string[], folderName: string): boolean {
  const target = folderName.toLowerCase() + '/';
  return files.some((file) => file.toLowerCase().startsWith(target));
}

function detectSegmentationFormat(content: string): boolean {
  // YOLO segmentation format has multiple coordinate pairs per line
  const lines = content.trim().split('\n');
  
  for (const line of lines) {
    const parts = line.trim().split(' ');
    // Segmentation has format: class_id x1 y1 x2 y2 ... xn yn
    // Detection has format: class_id x_center y_center width height (5 parts)
    if (parts.length > 5) {
      // Check if it looks like polygon points (even number of coordinates after class_id)
      const coordCount = parts.length - 1;
      if (coordCount % 2 === 0 && coordCount >= 6) {
        return true;
      }
    }
  }
  
  return false;
}

function extractPascalVOCClasses(zip: JSZip, files: string[]): string[] {
  const classes = new Set<string>();
  // This will be populated during parsing - for now just estimate
  return Array.from(classes);
}

File: src-tauri/src/import/format_detector.rs
use std::io::Read;
use zip::ZipArchive;

use super::DetectionResult;

pub fn detect(archive: &mut ZipArchive<std::fs::File>) -> Result<DetectionResult, String> {
    let file_names: Vec<String> = (0..archive.len())
        .filter_map(|i| archive.by_index(i).ok().map(|f| f.name().to_string()))
        .collect();
    let lower_files: Vec<String> = file_names.iter().map(|f| f.to_lowercase()).collect();

    // YOLO: classes.txt + data.yaml + labels/
    if has_file(&lower_files, "classes.txt") && has_file(&lower_files, "data.yaml") {
        if let Some(result) = detect_yolo(archive, &lower_files) {
            return Ok(result);
        }
    }

    // U-Net Masks: masks/ + images/
    if has_folder(&lower_files, "masks") && has_folder(&lower_files, "images") {
        if let Some(result) = detect_unet(&lower_files) {
            return Ok(result);
        }
    }

    // TIX: annotations.json con structure TIX (check before COCO)
    if has_file(&lower_files, "annotations.json") && has_folder(&lower_files, "images") {
        if let Some(result) = detect_tix(archive) {
            return Ok(result);
        }
    }

    // COCO: annotations.json with images/annotations/categories
    if has_file(&lower_files, "annotations.json") {
        if let Some(result) = detect_coco(archive) {
            return Ok(result);
        }
    }

    // Pascal VOC: Annotations/ + images/ (or JPEGImages/)
    if has_folder(&lower_files, "annotations") {
        if let Some(result) = detect_pascal_voc(&lower_files) {
            return Ok(result);
        }
    }

    // CSV: annotations.csv
    if has_file(&lower_files, "annotations.csv") {
        if let Some(result) = detect_csv(archive, &lower_files) {
            return Ok(result);
        }
    }

    // Folders by class: directories with images
    if let Some(result) = detect_folders_by_class(&lower_files) {
        return Ok(result);
    }

    Err("No se pudo detectar el formato del dataset".to_string())
}

fn has_file(files: &[String], name: &str) -> bool {
    let target = name.to_lowercase();
    files.iter().any(|f| {
        let f = f.to_lowercase();
        (f == target) || (f.ends_with(&format!("/{}", target)) && !f[..f.len() - target.len() - 1].contains('/'))
    })
}

fn has_folder(files: &[String], folder: &str) -> bool {
    let target = format!("{}/", folder.to_lowercase());
    files.iter().any(|f| f.starts_with(&target))
}

fn read_file_text(archive: &mut ZipArchive<std::fs::File>, name: &str) -> Option<String> {
    // Try exact name first, then case-insensitive
    for i in 0..archive.len() {
        if let Ok(mut file) = archive.by_index(i) {
            if file.name().eq_ignore_ascii_case(name) {
                let mut content = String::new();
                file.read_to_string(&mut content).ok()?;
                return Some(content);
            }
        }
    }
    None
}

fn detect_yolo(archive: &mut ZipArchive<std::fs::File>, files: &[String]) -> Option<DetectionResult> {
    let content = read_file_text(archive, "classes.txt")?;
    let class_count = content.trim().lines().filter(|l| !l.trim().is_empty()).count();

    // Check if segmentation by looking at a label file
    let txt_files: Vec<&String> = files.iter()
        .filter(|f| f.starts_with("labels/") && f.ends_with(".txt"))
        .collect();

    if txt_files.is_empty() {
        return None;
    }

    let label_content = read_file_text(archive, &txt_files[0])?;
    let is_segmentation = detect_segmentation_format(&label_content);

    Some(DetectionResult {
        format: if is_segmentation { "yolo-segmentation".to_string() } else { "yolo-detection".to_string() },
        project_type: if is_segmentation { "polygon".to_string() } else { "bbox".to_string() },
        confidence: 0.95,
        class_count: Some(class_count),
    })
}

fn detect_coco(archive: &mut ZipArchive<std::fs::File>) -> Option<DetectionResult> {
    let content = read_file_text(archive, "annotations.json")?;
    let data: serde_json::Value = serde_json::from_str(&content).ok()?;

    let annotations = data.get("annotations")?.as_array()?;
    let _images = data.get("images")?.as_array()?;
    let categories = data.get("categories")?.as_array()?;

    let has_segmentation = annotations.iter().any(|a| a.get("segmentation").is_some());
    let project_type = if has_segmentation { "instance-segmentation" } else { "bbox" };

    Some(DetectionResult {
        format: "coco".to_string(),
        project_type: project_type.to_string(),
        confidence: 0.95,
        class_count: Some(categories.len()),
    })
}

fn detect_tix(archive: &mut ZipArchive<std::fs::File>) -> Option<DetectionResult> {
    let content = read_file_text(archive, "annotations.json")?;
    let data: serde_json::Value = serde_json::from_str(&content).ok()?;

    // TIX has images array and optionally project.type
    let images = data.get("images")?.as_array()?;
    if images.is_empty() {
        return None;
    }

    // Check if it's COCO format (has annotations+categories arrays at root)
    if data.get("annotations").is_some() && data.get("categories").is_some() {
        return None; // This is COCO, not TIX
    }

    let project_type = data.get("project")
        .and_then(|p| p.get("type"))
        .and_then(|t| t.as_str())
        .map(|t| normalize_project_type(t))
        .unwrap_or_else(|| "bbox".to_string());

    let class_count = data.get("project")
        .and_then(|p| p.get("classes"))
        .and_then(|c| c.as_array())
        .map(|c| c.len());

    Some(DetectionResult {
        format: "tix".to_string(),
        project_type,
        confidence: 0.95,
        class_count,
    })
}

fn detect_pascal_voc(files: &[String]) -> Option<DetectionResult> {
    let xml_files: Vec<&String> = files.iter()
        .filter(|f| f.starts_with("annotations/") && f.ends_with(".xml"))
        .collect();

    if xml_files.is_empty() {
        return None;
    }

    Some(DetectionResult {
        format: "pascal-voc".to_string(),
        project_type: "bbox".to_string(),
        confidence: 0.9,
        class_count: None,
    })
}

fn detect_csv(archive: &mut ZipArchive<std::fs::File>, _files: &[String]) -> Option<DetectionResult> {
    let content = read_file_text(archive, "annotations.csv")?;
    let first_line = content.lines().next()?.to_lowercase();

    let class_count = read_file_text(archive, "classes.csv")
        .map(|c| c.trim().lines().count().saturating_sub(1));

    if first_line.contains("xmin") || first_line.contains("xmax") {
        return Some(DetectionResult {
            format: "csv-detection".to_string(),
            project_type: "bbox".to_string(),
            confidence: 0.9,
            class_count,
        });
    }

    if first_line.contains("keypoint") || first_line.contains("_visible") {
        return Some(DetectionResult {
            format: "csv-keypoints".to_string(),
            project_type: "keypoints".to_string(),
            confidence: 0.9,
            class_count,
        });
    }

    if first_line.contains("landmark") {
        return Some(DetectionResult {
            format: "csv-landmarks".to_string(),
            project_type: "landmarks".to_string(),
            confidence: 0.9,
            class_count,
        });
    }

    if first_line.contains("class") || first_line.contains("label") {
        return Some(DetectionResult {
            format: "csv-classification".to_string(),
            project_type: "classification".to_string(),
            confidence: 0.85,
            class_count,
        });
    }

    None
}

fn detect_unet(files: &[String]) -> Option<DetectionResult> {
    let mask_files: Vec<&String> = files.iter()
        .filter(|f| {
            f.starts_with("masks/") &&
            [".png", ".jpg", ".jpeg", ".bmp", ".webp"].iter().any(|ext| f.ends_with(ext))
        })
        .collect();

    if mask_files.is_empty() {
        return None;
    }

    Some(DetectionResult {
        format: "unet-masks".to_string(),
        project_type: "instance-segmentation".to_string(),
        confidence: 0.9,
        class_count: Some(2),
    })
}

fn detect_folders_by_class(files: &[String]) -> Option<DetectionResult> {
    let image_exts = [".jpg", ".jpeg", ".png", ".gif", ".bmp", ".webp"];
    let mut folders = std::collections::HashSet::new();

    for f in files {
        let parts: Vec<&str> = f.split('/').collect();
        if parts.len() == 2 {
            let filename = parts[1].to_lowercase();
            if image_exts.iter().any(|ext| filename.ends_with(ext)) {
                folders.insert(parts[0].to_string());
            }
        }
    }

    if folders.len() >= 2 {
        Some(DetectionResult {
            format: "folders-by-class".to_string(),
            project_type: "classification".to_string(),
            confidence: 0.85,
            class_count: Some(folders.len()),
        })
    } else {
        None
    }
}

fn detect_segmentation_format(content: &str) -> bool {
    for line in content.trim().lines() {
        let parts: Vec<&str> = line.trim().split_whitespace().collect();
        if parts.len() > 5 {
            let coord_count = parts.len() - 1;
            if coord_count % 2 == 0 && coord_count >= 6 {
                return true;
            }
        }
    }
    false
}

fn normalize_project_type(t: &str) -> String {
    match t.to_lowercase().as_str() {
        "detection" => "bbox".to_string(),
        "segmentation" => "mask".to_string(),
        "instanceseg" | "instancesegmentation" | "instance-segmentation" => "instance-segmentation".to_string(),
        "multilabel" | "multi-label-classification" => "multi-label-classification".to_string(),
        "keypoint" => "keypoints".to_string(),
        "landmark" => "landmarks".to_string(),
        other => other.to_string(),
    }
}
